src/auto-evo/Miche.cs

```diff
@@ -156,26 +177,25 @@ public bool InsertSpecies(Species species, Patch patch, Dictionary<Species, floa
             return true;
         }
 
-        // TODO: somehow avoid this recursive method call memory allocations
-        var newScores = new Dictionary<Species, float>();
+        var newScores = workingMemory.ScoreDictionaries[depth];
```

Comment:
Couldn't the cache be made so that this automatically increased the dictionary size if depth is higher? That way the dictionary wouldn't need to be pre-populated making it easier to use? Or could maybe even a stack (or maybe a queue) type be used instead of a dictionary? Or maybe just a plain list which is indexed by the depth? Dictionary lookups are more expensive than list index lookups.